src/objective/rank_objective.hpp

```diff
@@ -253,9 +277,41 @@ class LambdarankNDCG : public RankingObjective {
     }
   }
 
+  void UpdatePositionBiasFactors(const score_t* lambdas, const score_t* hessians) const override {
+    std::vector<double> bias_first_derivatives(num_position_ids_, 0.0);
+    std::vector<double> bias_second_derivatives(num_position_ids_, 0.0);
+    for (data_size_t i = 0; i < num_data_; i++) {
+      // accumilate first derivatives of utility w.r.t. position bias factors, for each position
+      bias_first_derivatives[positions_[i]] -= lambdas[i];
+      // accumilate second derivatives of utility w.r.t. position bias factors, for each position
+      bias_second_derivatives[positions_[i]] -= hessians[i];
+    }
+    #pragma omp parallel for schedule(guided)
+    for (data_size_t i = 0; i < num_position_ids_; i++) {
+      // do Newton-Rhapson step to update position bias factors
+      pos_biases_[i] += learning_rate_ * bias_first_derivatives[i] / (std::abs(bias_second_derivatives[i]) + 0.001);
+    }
+    LogDebugPositionBiasFactors();
```

Comment:
Shall this be included in a `#ifdef DEBUG` micro?

Reply:
It would be great to have a way to get the inferred position bias factors even in the release version (without checking #ifdef DEBUG). That's why I thought it would be easy to have them printed via Log::Debug() when we set parameter verbosity > 0. Ideally, we may want to output these factors into a file, but I'm not sure it is easy to implement.

Reply:
Now that the `LogDebugPositionBiasFactors` is using `Log::Debug`, I think it is OK to keep it without `#ifdef DEBUG`.